### DISCUSSION TITLE
Anyone having autocomplete working in Scala 3 projects?

### DISCUSSION THREAD

**Discussion by jan0sch** [scalameta/nvim-metals#451] [Q&A]
Hi,

I noticed that in my Scala 3 projects I don't have autocomplete via metals. With Scala 2 projects it works as usual and the Scala 3 ones do also have code navigation (jump to X) and so but no autocomplete. The default autocomplete of nvim/vim goes a long way, especially if a lot of files are in the buffers but for other stuff it is quite cumbersome.

Any ideas or something I am missing?

Kind regards,

Jens

P.S.:

config from init.lua:

```lua
local nvim_metals_group = vim.api.nvim_create_augroup("nvim-metals", { clear = true })
vim.api.nvim_create_autocmd("FileType", {
  pattern = { "scala", "sbt", "java" },
  callback = function()
    require("metals").initialize_or_attach({})
  end,
  group = nvim_metals_group,
})
metals_config = require("metals").bare_config()
metals_config.settings = {
  bloopSbtAlreadyInstalled = true,
  showImplicitArguments = true,
  showInferredType = true,
  excludedPackages = {
    "akka.actor.typed.javadsl",
    "com.github.swagger.akka.javadsl"
  }
}
metals_config.init_options.statusBarProvider = "on"

map("n", "<leader>ws", '<cmd>lua require"metals".hover_worksheet()<CR>')
map("n", "gf", "<cmd>lua vim.lsp.buf.definition()<CR>")
map("n", "gF", "<cmd>tab split | lua vim.lsp.buf.definition()<CR>")
map("n", "K", "<cmd>lua vim.lsp.buf.hover()<CR>")
map("n", "<leader>t", "<cmd>lua vim.lsp.buf.hover()<CR>")
map("n", "gi", "<cmd>lua vim.lsp.buf.implementation()<CR>")
map("n", "gr", "<cmd>lua vim.lsp.buf.references()<CR>")
map("n", "gds", "<cmd>lua vim.lsp.buf.document_symbol()<CR>")
map("n", "gws", "<cmd>lua vim.lsp.buf.workspace_symbol()<CR>")
map("n", "<leader>cl", [[<cmd>lua vim.lsp.codelens.run()<CR>]])
map("n", "<leader>sh", [[<cmd>lua vim.lsp.buf.signature_help()<CR>]])
map("n", "<leader>rn", "<cmd>lua vim.lsp.buf.rename()<CR>")
map("n", "<leader>f", "<cmd>lua vim.lsp.buf.formatting()<CR>")
map("n", "<leader>ca", "<cmd>lua vim.lsp.buf.code_action()<CR>")
map("n", "<leader>aa", [[<cmd>lua vim.diagnostic.setqflist()<CR>]]) -- all workspace diagnostics
map("n", "<leader>ae", [[<cmd>lua vim.diagnostic.setqflist({severity = "E"})<CR>]]) -- all workspace errors
map("n", "<leader>aw", [[<cmd>lua vim.diagnostic.setqflist({severity = "W"})<CR>]]) -- all workspace warnings
map("n", "<leader>d", "<cmd>lua vim.diagnostic.setloclist()<CR>") -- buffer diagnostics only
map("n", "<F2>", "<cmd>lua vim.diagnostic.goto_next { wrap = false }<CR>")
map("n", "<S-F2>", "<cmd>lua vim.diagnostic.goto_prev { wrap = false }<CR>")
```

Running the metals doctor gives the following output:

```text
## Metals Info
  - Metals Server version: 0.11.8
  - Build server currently being used is sbt v1.7.1.
  - Metals Java: 17.0.4 from OpenJDK BSD Porting Team located at /usr/local/openjdk17

These are the installed build targets for this workspace. One build target corresponds to one classpath. For example, normally one sbt project maps to two build targets: main and test.

## Build Targets

### comparison
  - target type: Scala 3.1.3
  - compilation: ✅ 
  - diagnostics: ✅ 
  - interactive: ✅ 
  - semanticdb: ✅ 
  - debugging: ✅ 
  - java: ⚠️ 

### comparison-it
  - target type: Scala 3.1.3
  - compilation: ✅ 
  - diagnostics: ✅ 
  - interactive: ✅ 
  - semanticdb: ✅ 
  - debugging: ✅ 
  - java: ⚠️ 

### comparison-test
  - target type: Scala 3.1.3
  - compilation: ✅ 
  - diagnostics: ✅ 
  - interactive: ✅ 
  - semanticdb: ✅ 
  - debugging: ✅ 
  - java: ⚠️ 

### comparison-build
  - target type: sbt 1.7.1
  - compilation: ✅ 
  - diagnostics: ⚠️ 
  - interactive: ✅ 
  - semanticdb: ✅ 
  - debugging: ❌
  - java: ⚠️ 
  - recommendation: Diagnostics and debugging for sbt are not supported currently.
```

**Comment by ckipp01** [comment]
Hey @jan0sch thanks for the report. Completions in Scala 3 should work the same as when on Scala 2. Nothing from your setup looks out of place, so it should work the same. If you do a `:MetalsToggleLogs` do you see anything funky? Also is this happening for you on a minimal project?

**Reply by jan0sch** [reply]
Logs look fine, just the regular status stuff and the compilation errors for wrong code.

Yes, even with starting a plain sbt project with scala 3 and minimal settings I never get autocomplete. I first thought that this is maybe "not there yet" for v3 because I didn't get any error messages like "autocomplete will not work" or so. A note: I use metals not with bloop but with sbt as a build server.

Settings for sbt:

```scala
inThisBuild(
  Seq(
    scalaVersion := "3.1.3",
    scalacOptions ++= Seq(
        "-deprecation",
        "-explain-types",
        "-feature",
        "-language:higherKinds",
        "-language:implicitConversions",
        "-unchecked",
        "-Xfatal-warnings",
        "-Ykind-projector",
    ),
    Compile / console / scalacOptions --= Seq("-Xfatal-warnings"),
    Test / console / scalacOptions --= Seq("-Xfatal-warnings")
  )
)
```

**Reply by ckipp01** [reply]
🤔 So I just tested this out by doing:

```
g8 scala/scala3.g8
```
Then replacing the `build.sbt` with what you have posted above, and then navigating to `Main.scala`. There I do have completions. What completion plugin or framework are you using?

**Reply by jan0sch** [reply]
Doesn't work for me. As for the autocompletion framework: I don't use any, just the plain stuff that comes with nvim. I expected that metals would be hooked into the ctrl+n/ctrl+p actions of nvim. Also there was nothing mentioned in the docs that one would need a "completion framework"

Besides that I use UltiSnips but that is mapped to ctrl+j and doesn't interfere with the regular completion.

**Reply by ckipp01** [reply]
No you don't _need_ one. Ins completion should work fine. Do you have `vim.lsp.omnifunc()` set up? If you're not using any completion framework you'll also want to make sure that it's set up. Somewhere you should have
```
  vim.api.nvim_buf_set_option(bufnr, 'omnifunc', 'v:lua.vim.lsp.omnifunc')
```
or something of the like.

**Reply by jan0sch** [reply]
That one I don't have, thanks for the info. 
However, I though that mixing nvim-lsp-config and nvim-metals is not suggested, am I right?
So where would I place this? So far I have only seen it inside an `on_attach` function of lsp-config blocks. :-?

**Reply by ckipp01** [reply]
> However, I though that mixing nvim-lsp-config and nvim-metals is not suggested, am I right?

Correct, but `vim.lsp.omnifunc` isn't part of `nvim-lspconfig`, it's part of Neovim core. You can follow the same pattern and put it inside of the `on_attach` of `nvim-metals`. For example you can see how I use `metals_config.on_attatch` [here](https://github.com/ckipp01/dots/blob/master/nvim/.config/nvim/lua/mesopotamia/lsp.lua#L65). So you'll want to add the call there.

**Reply by jan0sch** [reply]
I have added it into my config (many thanks by the way) but if I trigger completion I get the error "Option omnifunc not set" :-(

Here is my config, anything that directly jumps out?

```lua
local nvim_metals_group = vim.api.nvim_create_augroup("nvim-metals", { clear = true })
vim.api.nvim_create_autocmd("FileType", {
  pattern = { "scala", "sbt", "java" },
  callback = function()
    require("metals").initialize_or_attach({})
  end,
  group = nvim_metals_group,
})
metals_config = require("metals").bare_config()
metals_config.on_attach = function(client, bufnr)
  on_attach(client, bufnr)
  api.nvim_buf_set_option(bufnr, "omnifunc", "v:lua.vim.lsp.omnifunc")
end
metals_config.settings = {
  bloopSbtAlreadyInstalled = true,
  showImplicitArguments = true,
  showInferredType = true,
  excludedPackages = {
    "akka.actor.typed.javadsl",
    "com.github.swagger.akka.javadsl"
  }
}
metals_config.init_options.statusBarProvider = "on"
```

**Reply by jan0sch** [reply]
I played around with it some more and it seems that the `on_attach` function for metals isn't called at all, but I have no idea why. :-(

**Reply by ckipp01** [reply]
It looks like the issue is:

```lua
    require("metals").initialize_or_attach({})
```
When you are calling `initialize_or_attatch({})` you're calling it with an empty table instead of passing it `metals_config`.

**Reply by jan0sch** [reply]
Sadly this doesn't have any effect (config is still ignored). I moved the config definition before the autocommand and tried with and without `local` before the `metals_config` but to no avail.

The empty `{}` originated from the docs which state: "The empty `{}` that is passed into |initialize_or_attach()| is necessary and is very similar to the config object that is directly passed into the |vim.lsp.start_client()| method."

For other lsps I configured directly it works like charm.

```lua
local on_attach = function(client, bufnr)
  -- Vervollständigung via <ctrl-x><ctrl-o>
  vim.api.nvim_buf_set_option(bufnr, 'omnifunc', 'v:lua.vim.lsp.omnifunc')
end

require('lspconfig')['rust_analyzer'].setup{
  on_attach = on_attach,
  flags = lsp_flags,
  -- Server-specific settings...
  settings = {
    ["rust-analyzer"] = {}
  }
}
```

I'm tempted to kill my nvim config and start out from scratch but I haven't really time for that. :-(

**Reply by ckipp01** [reply]
> Sadly this doesn't have any effect (config is still ignored).

Your entire config? Even metals status isn't working?

>   bloopSbtAlreadyInstalled = true,

I'd recommend removing this unless you have a very good reason to keep it as this will prevent metals from adding the necessary `metals.sbt`.

**Reply by jan0sch** [reply]
> Your entire config? Even metals status isn't working?

The status bar stuff works.

> > bloopSbtAlreadyInstalled = true,
> 
> I'd recommend removing this unless you have a very good reason to keep it as this will prevent metals from adding the necessary `metals.sbt`.

I prefer to use sbt as backend so I disable the automatic bloop installation (which seems to be the unchangeable default) and use the `:MetalsSwitchBsp` command which will create the `metals.sbt` files.

**Reply by ckipp01** [reply]
> The status bar stuff works.

Ok, so then your config is being taken into account.

**Reply by jan0sch** [reply]
Weird, I just kicked out everything from init.lua except nvim-lspconfig, nvim-metals and plenary resulting in the behaviour staying the same (`Option 'omnifunc' not set`).

Maybe I have to nuke my whole nvim config folders and caches and start from scratch after all. :-(

Or I just try to fiddle with some "autocomplete plugin", any recommendations for a good one?

**Reply by ckipp01** [reply]
A lot of users (including myself) use https://github.com/hrsh7th/nvim-cmp.

**Reply by jan0sch** [reply]
Thanks, I gave it a try and indeed it includes the metals autocomplete. But I remembered why I have removed all the autocomplete plugins in the first place (despite of all the speed they feel slugish).

So I will try to make it work with default nvim completion. If I succeed I'll report my findings/solution here.

**Reply by jan0sch** [reply]
I was unable to get autocomplete working without nvim-cmp but at least I could configure it to only trigger completion upon key press (`c-n`): https://github.com/hrsh7th/nvim-cmp/discussions/1199

Thanks a bunch for your help and for nvim-metals of course. :-)
